File: LeetCode/Permutations.cpp
class Solution {
public:
    vector<vector<int> > permute(vector<int> &num) {
        // Add an empty vector as the base case (empty input)
        vector<vector<int> > permutations(1, vector<int>());
        // Algrithm description:
        //  Insert the current number in different spaces of previous permutations
        for (vector<int>::size_type index = 0; index != num.size(); ++index)
        {
            vector<vector<int> > subPermutations(permutations);
            permutations.clear();
            for (vector<vector<int> >::size_type i = 0; i != subPermutations.size(); ++i)
            {
                for (int offset = 0; offset != subPermutations[i].size()+1; ++offset)
                {
                    vector<int> temp(subPermutations[i]);
                    temp.insert(temp.begin() + offset, num[index]);
                    permutations.push_back(temp);
                }
            }
        }
        return permutations;
    }
};

File: LeetCode/Permutations.py
class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def permute(self, num):
        length = len(num)
        if length == 0: return []
        if length == 1: return [num]
        res = []
        for i in xrange(length):
            for j in self.permute(num[0:i] + num[i+1:]):
                res.append([num[i]] + j)
        return res

if __name__ == '__main__':
    s = Solution()
    print s.permute([1,1,2])